src/sentry/api/endpoints/organization_events.py

```diff
@@ -218,64 +232,155 @@ def get(self, request, organization):
+        snuba_args = deepcopy(snuba_args)
         data = raw_query(
             groupby=[key],
-            aggregations=snuba_args.get('aggregations', []) + [
+            conditions=snuba_args.pop('conditions', []) + [
+                [['isNotNull', [key]], '=', 1]
+            ],
+            aggregations=snuba_args.pop('aggregations', []) + [
                 ['count()', '', 'count'],
                 ['min', 'timestamp', 'first_seen'],
                 ['max', 'timestamp', 'last_seen'],
-                ['uniq', key, 'values_seen'],
             ],
             orderby='-count',
             referrer='api.organization-events-heatmap',
             **snuba_args
-        )['data']
-        return data
-
-    def _create_tag_key_tag_value_objects(self, key, data):
-        tag_values = []
-        values_seen = 0
-        for datum in data:
-            tag_values.append(TagValue(
-                key=key,
-                value=datum['value'],
-                times_seen=datum['count'],
-                last_seen=datum['last_seen'],
-                first_seen=datum['first_seen'],
-            ))
-            values_seen += datum['count']
-
-        tag_key = TagKey(
-            key=key,
-            values_seen=len(tag_values),
-            count=values_seen,
-            top_values=tag_values,
         )
-        return tag_key
+        return data['data']
+
+    def _create_tag_objects(self, keys, total_count, top_values):
+        tag_keys_dict = OrderedDict()
+
+        for top_value in top_values:
+            key = top_value['tags_key']
+
+            if key not in tag_keys_dict:
+                tag_keys_dict[key] = TagKey(
+                    key=key,
+                    top_values=[],
+                    count=total_count,
+                )
+            tag_keys_dict[key].top_values.append(
+                TagValue(
+                    key=key,
+                    value=top_value['tags_value'],
+                    times_seen=top_value['count'],
+                    first_seen=top_value['first_seen'],
+                    last_seen=top_value['last_seen'],
+                )
+            )
+
+        # Add categories with no values
+        for key in keys:
+            if key not in tag_keys_dict:
+                tag_keys_dict[key] = TagKey(
+                    key=key,
+                    top_values=[],
+                    count=total_count,
+                )
+        return tag_keys_dict.values()
+
+    def get_total_value_count(self, **kwargs):
+        kwargs = deepcopy(kwargs)
+        aggregations = kwargs.pop('aggregations', [])
+        aggregations += [
+            ['count()', '', 'count'],
+        ]
+
+        total_count = raw_query(
+            aggregations=aggregations,
+            referrer='api.organization-events-heatmap',
+            **kwargs
+        )['data'][0]['count']
+        return total_count
+
+    def get_top_values(self, keys, value_limit=TOP_VALUES_DEFAULT_LIMIT, **kwargs):
```

Comment:
This endpoint is pretty beefy. Should we consider moving some of the tag fetching logic into another module of functions? Something like `src/sentry/tagstore/logic.py`?

Reply:
Ok @markstory let me know if the changes I made were what you were thinking here.